store/postgres/src/store.rs

```diff
@@ -405,7 +412,45 @@ impl Store {
         data: Entity,
         event_source: EventSource,
     ) -> Result<(), StoreError> {
-        use db_schema::entities;
+        use db_schema::entities::{self, dsl};
+
+        // Collect all types that share an interface implementation with this
+        // entity type, and make sure there are no conflicting IDs.
+        //
+        // To understand why this is necessary, suppose that `Dog` and `Cat` are
+        // types and both implement an interface `Pet`, and both have instances
+        // with `id: "Fred"`. If a type `PetOwner` has a field `pets: [Pet]`
+        // then with the value `pets: ["Fred"]`, there's no way to disambiguate
+        // if that's Fred the Dog, Fred the Cat or both.
```

Comment:
It's worth noting (and should probably go into the comment) that this check only works as long as there is one writer per subgraph. If we ever had multiple concurrent writers, they could race against each other here, and successfully insert a cat and a dog with the same id.

Reply:
That's true, I've added this to the comment. If/when we get parallel writes from different data sources, we'll need to remember to adjust this.